File: zig/src/problem0003/longest_substring_without_repeating_characters.zig
const std = @import("std");

pub fn length_of_longestSubstring(s: []const u8) i32 {
    var hash_table = std.AutoHashMap(u8, i32).init(std.heap.page_allocator);
    defer hash_table.deinit();

    var left: i32 = 0;
    var right: i32 = 0;
    var res: i32 = 0;

    while (right < s.len) {
        const char = s[@intCast(right)];
        // * 如果哈希表中存在重复字符，则更新滑动窗口的左边界到重复字符的下一个位置
        if (hash_table.contains(char) and hash_table.get(char).? >= left) {
            left = hash_table.get(char).? + 1;
        }

        // * 将当前字符存入哈希表
        hash_table.put(char, right) catch unreachable;
        // * 更新滑动窗口的右边界
        right += 1;

        res = @max(res, right - left);
    }

    return res;
}

File: typescript/src/problem0003/longestSubstringWithoutRepeatingCharacters.ts
export default function lengthOfLongestSubstring(s: string): number {
  const hashTable = new Map<string, number>();
  let left = 0,
    right = 0,
    res = 0;

  while (right < s.length) {
    const char = s[right];
    // * 如果哈希表中存在重复字符，则更新滑动窗口的左边界到重复字符的下一个位置
    if (hashTable.has(char) && hashTable.get(char)! >= left) {
      left = hashTable.get(char)! + 1;
    }

    // * 将当前字符存入哈希表，并更新滑动窗口的右边界
    hashTable.set(char, right++);
    // * 更新最长无重复子串的长度
    res = Math.max(res, right - left);
  }

  return res;
}
